docs/async-use-cases.md

```diff
@@ -0,0 +1,95 @@
+may support it, but it's definitely not going to work on Desktop JS.  I think we should consider
+this invalid.
+
+## CallbackInterface / Foreign thread / Same thread / Blocking
+
+This is a normal CallbackInterace call and is already supported.
+
+## CallbackInterface / Foreign thread / Same thread / Async
+
+Async call from Rust to the foreign language.  Some use cases:
+   - Nimbus JEXL evaluation
+   - Async viaduct call
+
+Since this is happening on a foreign thead, it means there's a foreign caller that's waiting for a
+return.  There's a couple ways to handle that:
+
+  - Rust returns void, which is the current plan for Nimbus.  This doesn't mean it's a no-op, since
+    when the async calls resolve the Rust code will continue to do work.
+  - The Rust call is async, so it returns a Future/Promise (this is the other side of the
+    `async Nimbus.init()` use case described above).
+
+## CallbackInterface / Foreign thread / Thread switch / * and CallbackInterface / Rust Thread / Same thread / *
+
+This is invalid because it's not safe for foreign code to run on a Rust thread (see above).
+
+## CallbackInterface / Rust Thread / Thread switch / Blocking
+
+A Rust thread wants to make a blocking call into the foreign code, which requires a thread switch.
+For example, Rust creates a thread pool of workers.  Sometimes those workers want to make network
+calls, so they make a viaduct call.  Since this is happening in a thread pool, it's fine for the
+thread to block on the result.
+
+## CallbackInterface / Rust Thread / Thread switch / Async
+
+This is the same as the last one, except the call is async.  The main use case is probably the same
+as the previous section, except when Rust is using an event loop rather than a thread pool.
+
+# New features
+
+Based on the above analysis, I believe there are 2 features that we should consider adding:
+
+  - Rust code making an async call into a CallbackInterface / Foreign code
+    making an async into Rust. I think this could be implemented by having
+    UniFFI generate something like the hand-written demo from #1252.
+  - Handling CallbackInterface calls from a Rust-based thread. I'm not exactly sure how this would work, there's at least 2 options here:
+    - Handle everything on the foreign side. In the generate code we currently register a callback to invoke a CallbackInterface call. We might be able to update that code so that it schedules the call to run on the correct thread. But this assumes that it's safe to call that callback on the Rust thread. Is that true for all of our current languages? Are we okay with adding this requirement for future languages?
+    - Use a queue plus a waker. Push the CallbackInterface call to a queue, signal the foreign side (maybe writing a byte to a pipe or socket), then the foreign side would wake up and try to read from the queue. This seems more complicated than the first system, but might let us support more foreign languages.
```

Comment:
This feature seems necessary if we ever want to make a call from a Rust controlled (or C++ controlled) thread into JS.  Maybe we want that for the Nimubs dekstop work, but I'm not sure.